feat: add vertical timeline cursor

jeffpamer opened the pull request:
I would consider this a first iteration of the timeline cursor, with only the current mouse position tracked. We should take a second pass once the more thorough timeline UI overhaul is completed to match the new styles. And also at a later point we can add the pin icons at the same time we introduce the pinned vertical guides functionality.

![2022-09-14 16 22 20](https://user-images.githubusercontent.com/888818/190280850-b7dd1d59-a08d-43c4-8513-c1aef07dfffb.gif)

jeffpamer committed:
prototype rendering cursor underneath timeline elements

jeffpamer committed:
position cursor using transform instead of left

jeffpamer committed:
adjust x axis tick visibility so cursor is visible

jeffpamer committed:
add DOY time label to cursor header

jeffpamer committed:
move timeline back out of cursor hierarchy

AaronPlave commented:
This is so nice! One comment/question – any thoughts about disabling the vertical hover cursor when dragging an activity? Wondering if things are too busy with everything displaying together in that area and the activity does have the time in the tooltip which.. mostly shows up when dragging. But also maybe this is something to tackle when all of this gets reworked for the new designs.
<img width="371" alt="image" src="https://user-images.githubusercontent.com/4419607/190284353-1196cb2d-3f7e-453c-96d1-c1827d2cf42e.png">

camargo commented:
It looks like the last timeline row is being occluded at about the same height of the new cursor row. We need to account for the row container max height with respect to the new cursor row height. I *think* we can just add the height of the new cursor row to this function so the correct "rows max height" is calculated: https://github.com/NASA-AMMOS/aerie-ui/blob/develop/src/components/timeline/Timeline.svelte#L103

<img width="1013" alt="Screen Shot 2022-09-15 at 8 18 08 AM" src="https://user-images.githubusercontent.com/683355/190442358-be58d8fd-e457-4634-9397-b1acdd52a405.png">

camargo commented:
Can we add a timeline view configuration and hot-key toggle to turn the cursor on and off?

jeffpamer commented:
All great observations and ideas!

- I'll patch up the height adjustment needed to fit the cursors and the last row.
- I like the idea of hiding the cursor on dragging. Though I do see the value of being able to easily see the time you're dragging _to_. But maybe that's covered by the info in the tooltip?
- Love the idea of a hot-key toggle. Do you think we want to default to on? Are you thinking we'd have a new icon/button up in the header to toggle as well?

camargo commented:
> Love the idea of a hot-key toggle. Do you think we want to default to on? Are you thinking we'd have a new icon/button up in the header to toggle as well?

Yes default should be on. And for now just a hot-key toggle on/off. If users want we can add an icon to the header next iteration, but we are trying to not make the header super busy with toggles right now.

jeffpamer committed:
Merge remote-tracking branch 'origin' into feature/vertical-timeline-cursor

jeffpamer commented:
I think for now I'm leaning towards leaving the vertical cursor on while dragging activities. Currently the tooltip appears when you're actively hovering over the activity, but when dragging you can move above/below so it disappears. So in that case it's nice the cursor still is present. Certainly open to other opinions here, but maybe we can just try it out for a while and then revisit as needed.

I'll push up the change to fix the height and add the toggle hot-key today.

jeffpamer committed:
set cursor header height by prop, reflect that size in total usable height for rows

jeffpamer committed:
use static references to elements instead of reselecting with d3 each time. place label using transforms

jeffpamer committed:
show/hide cursor based on hot-key

jeffpamer commented:
- Corrected the height measurement so that we're now accounting for the height of the cursor header and the bottom row isn't cut off
- Added the hot-key (Ctrl+T) to toggle the cursor on/off
- Added a few small optimizations, like storing references to divs instead of reselecting them with D3 on every mouse move, and positioning the timestamp label using `transform` instead of `left`

I think this is all set for re-review!

camargo commented on src/components/timeline/Timeline.svelte:
```diff
@@ -43,6 +46,14 @@
   });
 
+  onMount(() => {
+    document.addEventListener('keydown', onKeyDown);
```
Instead of the `onMount` event listener bindings can we add:

```svelte
<svelte:window on:keydown={onKeydown} />
```

This should automatically add/remove the key listener from this component.

Ref: https://svelte.dev/tutorial/svelte-window

jeffpamer replied:
Oh nice I didn't know about that, that's much cleaner.

camargo commented on src/components/timeline/Timeline.svelte:
```diff
@@ -27,6 +28,8 @@
   let timelineDiv: HTMLDivElement;
   let xAxisDiv: HTMLDivElement;
   let xAxisDrawHeight: number = 90;
+  let cursorHeaderHeight: number = 20;
```
Alphabetize the `let`s

camargo commented on src/components/timeline/TimelineCursors.svelte:
```diff
@@ -0,0 +1,133 @@
+  export let mouseOver: MouseOver;
+  export let xScaleView: ScaleTime<number, number> | null = null;
+  export let marginLeft: number = 50;
+  export let drawWidth: number = 0;
```
Alphabetize `export`, `let`, and function names

camargo commented on src/components/timeline/TimelineCursors.svelte:
```diff
@@ -0,0 +1,133 @@
+    }
+  }
+
+  function onCursorEnableChange(cursorEnabled) {
```
Add type to the `cursorEnabled` param

camargo commented on src/components/timeline/TimelineCursors.svelte:
```diff
@@ -0,0 +1,133 @@
+      const doyTime = getDoyTime(new Date(unixEpochTime));
+
+      cancelAnimationFrame(raf);
+      raf = window.requestAnimationFrame(() => {
```
Is the `requestAnimationFrame` necessary here? My reading of this is the cursor should be updated on each `mouseOver`, which gets called on mouse pointer change.

jeffpamer replied:
It's maybe not "necessary", but improves FPS. `mouseOver` events will be fired async, and it's unknown at which point in the render cycle we'll respond to it. Theoretically multiple `mouseOver` events could happen in the span of one frame. Using `rAF` ensures we only tell the browser to update the cursor position when it's ready to paint, and makes sure we only update once per frame.

camargo replied:
Ah nice good to know. So you're optimizing the update. I *think* Svelte will take care of this since it only updates during the next microtask. So changes to `mouseOver` will be batched and the update will only happen once during the next microtask. See: https://lihautan.com/compile-svelte-in-your-head-part-2/#schedule-update
I'm fine with keeping this for now tho since I don't think it hurts anything.

jeffpamer committed:
Merge remote-tracking branch 'origin/develop' into feature/vertical-timeline-cursor

jeffpamer committed:
address pr formatting issues and use svelte:window instead of manually binding/unbinding keydown event

camargo reviewed (approved):
Tested locally and works great! Ready to merge
